File: RequestPermissionSample/app/src/main/java/com/example/app/MainActivity.java
package com.example.app;

import android.Manifest;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;

import com.example.app.utils.MYActionCallback;

public class MainActivity extends AppCompatActivity {

    private static final String CHANNEL_ID = "my-channel-123";
    private static final int ID = 123;
    private static final String TAG = "MainActivity";

    // Register the permissions callback, which handles the user's response to the
    // system permissions dialog. Save the return value, an instance of
    // ActivityResultLauncher, as an instance variable.
    private ActivityResultLauncher<String> requestPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
                if (isGranted) {
                    // Permission is granted. Continue the action or workflow in your
                    // app.
                    Log.i(TAG, "Granted: ");
                } else {
                    // Explain to the user that the feature is unavailable because the
                    // features requires a permission that the user has denied. At the
                    // same time, respect the user's decision. Don't link to system
                    // settings in an effort to convince the user to change their
                    // decision.
                    Log.i(TAG, "Not Granted: ");
                }
            });


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                NotificationCompat.Builder builder = new NotificationCompat.Builder(MainActivity.this, CHANNEL_ID)
                        .setSmallIcon(R.drawable.ic_launcher_background)
                        .setContentTitle("My notification title")
                        .setContentText("My notification text")
                        .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                        .setAutoCancel(true);

                NotificationManagerCompat notificationManager = NotificationManagerCompat.from(MainActivity.this);

                // notificationId is a unique int for each notification that you must define
                notificationManager.notify(ID, builder.build());
            }
        });

        createNotificationChannel();

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) == PackageManager.PERMISSION_GRANTED) {
            // You can use the API that requires the permission.
            Log.i(TAG, "Has permission");
        }  else {
            // You can directly ask for the permission.
            // The registered ActivityResultCallback gets the result of this request.
            requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS);
        }

    }

    @Override
    protected void onResume() {
        super.onResume();

        findViewById(R.id.button).startActionMode(new MYActionCallback());
    }

    private void createNotificationChannel() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            CharSequence name = "My notification channel name";
            String description = "My notification channel description";
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            NotificationChannel channel = new NotificationChannel(CHANNEL_ID, name, importance);
            channel.setDescription(description);
            // Register the channel with the system; you can't change the importance
            // or other notification behaviors after this
            NotificationManager notificationManager = getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
    }

}

File: Android14TestApp/app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.content.ComponentName
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import com.example.myapplication.ActivityLaunchHelper.launchFromBackgroundUsingPI

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.openActivityBtn).setOnClickListener {

            ActivityLaunchHelper.launchActivityFromApp(this)

            Log.i(TAG, "onCreate: called start activity")
        }

        findViewById<Button>(R.id.openActivityUsingPIBtn).setOnClickListener {

            ActivityLaunchHelper.launchActivityFromAppUsingPI(this)

            Log.i(TAG, "onCreate: called start activity")
        }

        findViewById<Button>(R.id.openServiceId).setOnClickListener {

            val serviceConnection = object : ServiceConnection {

                override fun onServiceConnected(p0: ComponentName?, p1: IBinder?) {
                    Log.i(TAG, "onServiceConnected $p0")
                }

                override fun onServiceDisconnected(p0: ComponentName?) {
                    Log.i(TAG, "onServiceDisconnected $p0")
                }
            }

            ActivityLaunchHelper.bindServiceFromApp(this, serviceConnection)

            Log.i(TAG, "onCreate: called start service")
        }

        Log.i(TAG, "onCreate: out")
    }

    override fun onResume() {
        super.onResume()
        Log.i(TAG, "onResume: out")
    }

    override fun onStop() {
        super.onStop()
        Log.i(TAG, "onStop: in")

        if (!IS_INTENT_SENT) {

            Handler(Looper.getMainLooper()).postDelayed({

                launchFromBackgroundUsingPI(MainActivity@this)

                //launchFromBackgroundUsingIntent()

            }, 10 * 1000)

            IS_INTENT_SENT = true;
        }

        Log.i(TAG, "onStop: out")
    }

    companion object {
        const val TAG = "APP_TEST1-MainActivity"
        var IS_INTENT_SENT = false
    }
}